clang/lib/AST/NormalizeUtils.cpp

```diff
@@ -0,0 +1,304 @@
+    return false;
+}
+
+// Input form:  E1 - E2
+// Output form: E1 + -E2
+Expr *NormalizeUtil::TransformSingleAdditiveOp(Sema &S, Expr *E) {
+  BinaryOperator *BO = dyn_cast<BinaryOperator>(E->IgnoreParens());
+  if (!BO)
+    return E;
+  if (BO->getOpcode() != BinaryOperatorKind::BO_Sub)
+    return E;
+
+  Expr *LHS = BO->getLHS();
+  Expr *RHS = BO->getRHS();
+  Expr *Minus = ExprCreatorUtil::CreateUnaryOperator(S, RHS,
+                                   UnaryOperatorKind::UO_Minus);
+  return AddExprs(S, LHS, Minus);
+}
+
+Expr *NormalizeUtil::AddExprs(Sema &S, Expr *LHS, Expr *RHS) {
+  return ExprCreatorUtil::CreateBinaryOperator(S, LHS, RHS,
+                            BinaryOperatorKind::BO_Add);
+}
+
+bool NormalizeUtil::AddConstants(llvm::APSInt &C1, llvm::APSInt C2) {
+  ExprUtil::EnsureEqualBitWidths(C1, C2);
+  bool Overflow;
+  C1 = C1.sadd_ov(C2, Overflow);
+  return !Overflow;
+}
+
+bool NormalizeUtil::GetRHSConstant(Sema &S, BinaryOperator *E, QualType T,
+                                   llvm::APSInt &Constant) {
+  if (!E->isAdditiveOp())
+    return false;
+  Optional<llvm::APSInt> OptConstant =
+                         E->getRHS()->getIntegerConstantExpr(S.Context);
+  if (!OptConstant)
+    return false;
+
+  Constant = *OptConstant;
+  bool Overflow;
+  Constant = ExprUtil::ConvertToSignedPointerWidth(S.Context, Constant, Overflow);
+  if (Overflow)
+    return false;
+  // Normalize the operation by negating the offset if necessary.
+  if (E->getOpcode() == BO_Sub) {
+    uint64_t PointerWidth = S.Context.getTargetInfo().getPointerWidth(0);
+    Constant = llvm::APSInt(PointerWidth, false).ssub_ov(Constant, Overflow);
+    if (Overflow)
+      return false;
+  }
+  llvm::APSInt ElemSize;
+  if (!ExprUtil::getReferentSizeInChars(S.Context, T, ElemSize))
+    return false;
+  Constant = Constant.smul_ov(ElemSize, Overflow);
+  if (Overflow)
+    return false;
+
+  return true;
+}
+
+bool NormalizeUtil::QueryPointerAdditiveConstant(Sema &S, Expr *E,
+                                                 Expr *&PointerExpr,
+                                                 llvm::APSInt &Constant) {
+  BinaryOperator *BO = dyn_cast<BinaryOperator>(E->IgnoreParens());
+  if (!BO)
+    return false;
+  if (!BO->isAdditiveOp())
+    return false;
+
+  // E must be of the form PointerExpr +/- Constant or Constant + PointerExpr,
+  // where PointerExpr has pointer type and Constant is an integer constant.
+  // Note that E cannot be of the form Constant - PointerExpr, since a pointer
+  // cannot appear on the right-hand side of a subtraction operator.
+  if (BO->getLHS()->getType()->isPointerType() &&
+      BO->getRHS()->isIntegerConstantExpr(S.Context))
+    PointerExpr = BO->getLHS();
+  else if (BO->getOpcode() == BinaryOperatorKind::BO_Add &&
+              BO->getRHS()->getType()->isPointerType() &&
+              BO->getLHS()->isIntegerConstantExpr(S.Context))
+    PointerExpr = BO->getRHS();
+  else
```

Comment:
I see you fixed the same build error that I encountered when I tried to merge `master` and `updated_baseline_master_12`.  But doesn't this need a `getIntegerConstantExpr` now that `isIntegerConstantExpr` no longer does that as a side effect?

```suggestion
  if (BO->getLHS()->getType()->isPointerType() &&
      BO->getRHS()->isIntegerConstantExpr(S.Context)) {
    PointerExpr = BO->getLHS();
    Constant = *BO->getRHS()->getIntegerConstantExpr(S.Context);
  } else if (BO->getOpcode() == BinaryOperatorKind::BO_Add &&
              BO->getRHS()->getType()->isPointerType() &&
              BO->getLHS()->isIntegerConstantExpr(S.Context)) {
    PointerExpr = BO->getRHS();
    Constant = *BO->getLHS()->getIntegerConstantExpr(S.Context);
  } else
```

As with the `NormalizeUtil::GetRHSConstant` bug, if this wasn't caught by the existing Checked C tests, you might want to add more tests.

Reply:
@mattmccutchen-cci , thanks! There were test case failures and I was looking into those.

At this point, all features that were planned for the current release are on `master`. So, now I will work on finalizing the merge and the release (the released Checked C compiler will be based on LLVM 12.0.1). If you are still having merge issues/test failures with `updated_baseline_master_12` or `updated_baseline_master_12_temp`, I propose that you wait until the release, so that you don't have to work with branches that are behind `master`.